manual/src/refman/extensions/tail_mod_cons.etex

```diff
@@ -0,0 +1,548 @@
+\begin{verbatim}
+let[@tail_mod_cons] rec flatten li =
+  match li with
+  | [] -> []
+  | xs :: xss ->
+    append_flatten xs xss
+\end{verbatim}
+
+The call to "append_flatten", which syntactically appears in tail
+position, gets transformed differently whether the function has
+a destination-passing-style version available -- depending on whether
+it is itself marked "[\@tail_mod_cons]":
+\begin{verbatim}
+(* if append_flatten_dps exists *)
+and flatten_dps li dst i =
+  match li with
+  | [] -> dst.i <- []
+  | xs :: xss ->
+    append_flatten_dps xs xss dst i
+
+(* if append_flatten_dps does not exist *)
+and rec flatten_dps li dst i =
+  match li with
+  | [] -> dst.i <- []
+  | xs :: xss ->
+    dst.i <- append_flatten xs xss
+\end{verbatim}
+If "append_flatten" does not have a destination-passing-style version,
+the call gets transformed to a non-tail call.
+
+\subsection{ss:limitations}{Current limitations}
+
+\paragraph{Local, first-order transformation} When a function gets
+transformed with tail-mod-cons, two definitions are generated, one
+providing a direct-style interface and one providing the
+destination-passing-style version. However, not all calls to this
+function in tail-modulo-cons position will use the
+destination-passing-style version and become tail calls:
+\begin{itemize}
+\item The transformation is local: only tail-mod-cons calls to "foo"
+  within the same compilation unit as "foo" become tail calls.
+\item The transformation is first-order: only direct calls that
+  a known tail-mod-cons functions become tail calls when in
+  tail-mod-cons position, never calls to function parameters.
+\end{itemize}
+
+Consier the call "Option.map foo x" for example: even if "foo" is
+called in tail-mod-cons position within the definition of
+"Option.map", that call will never become a tail call. (This would the
+case even if the call to "Option.map" was inside the "Option"
+module.)
+
+In general this limitation is not a problem for recursive functions:
+the first call from an outside module or a higher-order function will
+consume stack space, but further recursive calls in tail-mod-cons
+position will get optimized. For example if "List.map" is defined as
+a tail-mod-cons function, calls from outside the "List" module will
+not become tail calls when in tail positions, but the recursive calls
+within the definition of "List.map" are in tail-modulo-cons positions
+and do become tail calls: processing the first element of the list
+will consume stack space, but all further elements are handled in
+constant space.
+
+The limitations may be an issue in more complex situations where
+mutual recursion happens between functions, with some functions not
+marked tail-mod-cons, or defined across different modules, or called
+indirectly, for example thourgh function paramaters.
+
+\paragraph{Non-exact calls to tupled functions} OCaml performs an
+implicit optimization for "tupled" functions, which take a single
```

Comment:
Are "tupled" functions the same as "uncurried" functions?

Reply:
Probably so, I'm not sure; the compiler uses "tupled" internally so I went for this naming. ("uncurried" is not used in the manual.)

Reply:
Thanks!